### ISSUE TITLE
python syntax file, support f-string

### ISSUE THREAD

**Issue by A4-Tacks** [vim/vim#14033]
Vim syntax file for Python, with support for f-string added
- f-string support
- Fixed shortstring with multiple lines still highlighted
- Update string prefix to Python3.10+ syntax

have contacted the maintainer zpetkovic@acm.org But there seems to be no response

**python.vim.patch**
```diff
--- /usr/share/vim/vim91/syntax/python.vim	2024-01-03 12:24:01.000000000 +0800
+++ python.vim	2024-02-14 12:03:07.910000000 +0800
@@ -143,17 +143,35 @@
 
 " Triple-quoted strings can contain doctests.
 syn region  pythonString matchgroup=pythonQuotes
-      \ start=+[uU]\=\z(['"]\)+ end="\z1" skip="\\\\\|\\\z1"
+      \ start=+[uUbB]\=\z(['"]\)+ end="$\|\z1" skip="\\\\\|\\\z1"
       \ contains=pythonEscape,@Spell
 syn region  pythonString matchgroup=pythonTripleQuotes
-      \ start=+[uU]\=\z('''\|"""\)+ end="\z1" keepend
+      \ start=+[uUbB]\=\z('''\|"""\)+ end="\z1" keepend
       \ contains=pythonEscape,pythonSpaceError,pythonDoctest,@Spell
 syn region  pythonRawString matchgroup=pythonQuotes
-      \ start=+[uU]\=[rR]\z(['"]\)+ end="\z1" skip="\\\\\|\\\z1"
+      \ start=+\%([bB][rR]\=\|[rR][bB]\=\)\z(['"]\)+
+      \ end="$\|\z1" skip="\\\\\|\\\r\=$\|\\\z1"
       \ contains=@Spell
 syn region  pythonRawString matchgroup=pythonTripleQuotes
-      \ start=+[uU]\=[rR]\z('''\|"""\)+ end="\z1" keepend
+      \ start=+\%([bB][rR]\=\|[rR][bB]\=\)\z('''\|"""\)+
+      \ end="\z1" keepend
       \ contains=pythonSpaceError,pythonDoctest,@Spell
+syn region  pythonFString matchgroup=pythonQuotes
+      \ start=+[fF]\z(['"]\)+
+      \ end="$\|\z1" skip="\\\\\|\\\z1"
+      \ contains=pythonEscape,pythonFStringEscapedBrace,pythonFStringValue,@Spell
+syn region  pythonFString matchgroup=pythonTripleQuotes
+      \ start=+[fF]\z('''\|"""\)+
+      \ end="\z1" keepend
+      \ contains=pythonEscape,pythonSpaceError,pythonFStringEscapedBrace,pythonFStringValue,pythonDoctest,@Spell
+syn region  pythonRawFString matchgroup=pythonQuotes
+      \ start=+\%([fF][rR]\|[rR][fF]\)\z(['"]\)+
+      \ end="$\|\z1" skip="\\\\\|\\\r\=$\|\\\z1"
+      \ contains=pythonFStringEscapedBrace,pythonFStringValue,@Spell
+syn region  pythonRawFString matchgroup=pythonTripleQuotes
+      \ start=+\%([fF][rR]\|[rR][fF]\)\z('''\|"""\)+
+      \ end="\z1" keepend
+      \ contains=pythonSpaceError,pythonFStringEscapedBrace,pythonFStringValue,pythonDoctest,@Spell
 
 syn match   pythonEscape	+\\[abfnrtv'"\\]+ contained
 syn match   pythonEscape	"\\\o\{1,3}" contained
@@ -161,7 +179,11 @@
 syn match   pythonEscape	"\%(\\u\x\{4}\|\\U\x\{8}\)" contained
 " Python allows case-insensitive Unicode IDs: http://www.unicode.org/charts/
 syn match   pythonEscape	"\\N{\a\+\%(\s\a\+\)*}" contained
-syn match   pythonEscape	"\\$"
+syn match   pythonEscape	"\\\r\=$"
+
+syn region  pythonFStringValue matchgroup=pythonFStringBrace
+      \ start=+{+ end=+=\=\%(![rsa]\)\=\%(:[^{}]*\%({[^{}]*\%({[^{}]*}[^{}]*\)*}[^{}]*\)*\)\=}+ contains=TOP contained excludenl
+syn match   pythonFStringEscapedBrace	"{{\|}}" contained
 
 " It is very important to understand all details before changing the
 " regular expressions below or their order.
@@ -312,9 +334,14 @@
 hi def link pythonTodo			Todo
 hi def link pythonString		String
 hi def link pythonRawString		String
+hi def link pythonFString		String
+hi def link pythonRawFString		String
 hi def link pythonQuotes		String
 hi def link pythonTripleQuotes		pythonQuotes
 hi def link pythonEscape		Special
+hi def link pythonFStringEscapedBrace	Special
+hi def link pythonFStringValue		Normal
+hi def link pythonFStringBrace		Include
 if !exists("python_no_number_highlight")
   hi def link pythonNumber		Number
 endif
```

**Comment by A4-Tacks** [comment]
`f"result: {value:{width}.{precision}}"` seems to be unable to be handled well

**Comment by A4-Tacks** [comment]
> `f"result: {value:{width}.{precision}}"` seems to be unable to be handled well

Modified, hardcoded several layers, maybe enough

**Comment by A4-Tacks** [comment]
> Thanks. Could you please create a PR?
> 
> > `f"result: {value:{width}.{precision}}"` seems to be unable to be handled well
> 
> You can add the `pythonFStringValue` group to its own `contains` list to simplify matching these nested replacement fields. Then the end pattern can probably be reduced to a closing brace.
> 
> Ping @zvezdan

I cannot add it to its own contains list, it should only take effect on the right side of the colon, and the expression needs to be highlighted on the left side

**Comment by zvezdan** [comment]
> have contacted the maintainer (...) But there seems to be no response

@A4-Tacks I found four emails from you in the spam quarantine of my email provider because they were ranked as spam at high-level of probability, due to the following _facts_:

* several of them sent successively immediately one after another -- raises suspicion
* they had an attachment encoded in base64 with rarely used extension (`.patch`) and type `text/x-diff` -- could have sent them as plain text instead
* sender's name `A4-Tacks` sounds like a word `Attacks` -- why this choice of name/username/nickname?
* sender's email address username part is a seemingly random selection of letters and numbers -- why the need to hide the identity?
* the sender's domain 163.com was suspected as a possible source of spam

Each of these issues individually perhaps wouldn't be a problem, but taken together they resulted in your email messages being quarantined by my provider. I review quarantined messages once a week and I see them now. I allowed them to be delivered and _carefully_ opened them (because I didn't trust them either) as _raw messages_, copied the base64 encoded content of the attachment into a file and then decoded it into an output file. It is indeed the patch similar to the one you have here.

**As you can see, I did not ignore your messages**, but, rather, you caused the issue by your choice to send the messages as shown above.

**Lessons learned**:

* My identity is not hidden and if you're sending me a personal message you should not hide your identity either. Our communication remains between us only. It also helps with giving you credit later.
* If you're sending patches, they're text and you can send them as plain text, to avoid spam filters ranking them.
* If the message with attachment didn't receive response, re-sending it repeatedly is not going to help. Assume the spam filter got activated and send a simple message to start the communication. Once we exchange a few messages, the spam filter will also act differently.
* Creating a GitHub issue or PR and tagging me on it as @dkearns did here is perhaps the simplest way and avoids all email spam filter issues. 🙂

**Comment by zvezdan** [comment]
> Ping @zvezdan

Thank you both @dkearns and @A4-Tacks!

I'm currently testing this change, but also a few other options that I worked on and some other people sent. Some of the options are whether the bytes should be treated separately from strings, allowing for separate highlight coloring for people who prefer that, etc. This patch/issue is not making that distinction. Also, getting the `contains` things right can be tricky. I have an implementation of similar handling already for docstrings and my personal reStructuredText syntax highlighting setup and I'm trying to get the best combination of options and functionality.

In the meantime, I'm inclined to send a simpler PR that would just support f-strings correctly without trying to do any interpretation inside them. That would resolve some other already open issues, as well as partially covering this one. Then, I am testing a setting similar to `python_no_doctest_highlight` that would allow switching ON/OFF the  highlighting and interpretation of the content of f-strings. I believe that such a change would add more flexibility for users with different preferences.

In any case, I am working on this change and will incorporate all valid ideas including the ones from this issue and the corresponding patch. Thank you again and I'll keep you posted about the results of my testing soon.

**Comment by A4-Tacks** [comment]
> I had a quick attempt at this on the back of @A4-Tacks' PR. This patch allows for replacement fields in the format spec and avoids premature termination of the expression by spurious matches for the conversion and format spec delimiters.
> 
> IIUC, the following are correctly matched:
> 
> ```python
> # Lib/fractions.py
> digits = f"{significand:0{point_pos + 1}d}"
> # Lib/test/test_fstring.py
> self.assertEqual(f"sadsd {1 + 1 =  :{1 + 1:1d}f}", "sadsd 1 + 1 =  2.000000")
> # example from PEP - 701
> f'Useless use of lambdas: { (lambda x: x*2) }'
> ```
> 

Why is the `def` abnormal after this? Did I apply the damaged patch?

![image](https://github.com/vim/vim/assets/102709083/7d63f6b8-3ad8-4146-864d-5bb577a057fa)
```python
# Lib/fractions.py
digits = f"{significand:0{point_pos + 1}d}"
# Lib/test/test_fstring.py
self.assertEqual(f"sadsd {1 + 1 =  :{1 + 1:1d}f}", "sadsd 1 + 1 =  2.000000")
# example from PEP - 701
f'Useless use of lambdas: { (lambda x: x*2) }'

def foo() -> None:
    pass
```

**Comment by zvezdan** [comment]
@dkearns The places where `ALLBUT` is used were added because they were needed to correctly display syntax in _doctests_, _matrix multiplication_, and when an object _attribute_ name matches a _builtin_ name. Any of these `ALLBUT` lists can and should be extended if needed when adding new features, such as f-strings. Of course, we can also have a separate conversation about the use `ALLBUT` in general and possible better solutions for the existing usage.

However, the PR associated with this issue is trying to do two things: add f-string support and change a definition of strings to include `bytes`. I prefer separate changes for separate purpose that can be also separately reverted if needed. Adding a full support for f-strings with all the correct `contains` lists is tricky by itself and, as I mentioned earlier, might be best done iteratively in separate changes. First a simple support for f-strings to be correctly highlighted as strings. Then more complex `contains` features.

There are other contributions and options that I'm looking into. Some of them prefer `bytes` to be a separate highlight category. People also prefer having an option to switch features ON/OFF.  I doubt that any of my own or contributed changes, including this PR, will go in directly as is into the syntax file. A combination of them, that I'm working on, sent as smaller separate commits that add features and functionality with increasing usefulness and with switches for their use will be probably the result of this work.

Give me a couple of days to prepare the changes please. Then we can start closing the outstanding issues and PRs as we commit them.

Again, I truly appreciate all the contributions and ideas for improvements from all of you!

**Comment by jgb** [comment]
Any updates on this one? Would be lovely to have good support for f-string syntax! 
I personally still use Vimjas/vim-python-pep8-indent to fix indentation and vim-python/python-syntax for better syntax. It would be great if both syntax and indentation could be brought up to speed on modern python standards.

**Comment by zvezdan** [comment]
I previously stated what I feel should be changed / improved in the [comment](https://github.com/vim/vim/issues/14033#issuecomment-1953414989) above.

Several people sent patches and I had a few candidates in my local changes for the syntax file.
I'll try to send a sequence of incremental improvements to the syntax file to bring it up-to-date.

**Comment by f4nb0y** [comment]
Hi @zvezdan,

I've sent #17548 as a possible first step for this.

1. Does the scope address your previous comments, or should the `pythonFStringField` changes be in a separate PR?

2. Do you have syntax tests that we can use for regression testing?

**Comment by f4nb0y** [comment]
Hi @zvezdan,

Do you have time to review #17548 ?

**Comment by zvezdan** [comment]
> Hi [@zvezdan](https://github.com/zvezdan),
> 
> Do you have time to review [#17548](https://github.com/vim/vim/pull/17548) ?

Next weekend. I’m writing this from a phone and won’t have a computer available until then.

**Comment by f4nb0y** [comment]
Hi Folks,

I've closed #17548 in favour of a more structured approach, to support all of the literal string syntaxes in Python 2.7 and 3.*.

I'll send a series of smaller PRs as follows:

- Python 3 bytes and raw bytes
    * Restrict Unicode escapes to strings
- Python 2 b-strings (allowed but ignored in 2.7)
- Python 2 u-strings
    * Restrict Unicode escapes to u-strings
- Python 3 f-strings
- Python 3 f-string fields
- Python 3 printf string formatting
- Python 2 printf string formatting

We also need tests to ensure that we support as many cases as possible, and existing highlighting is not impacted. @dkearns added a large test to #14057, but some of that appears to be copied from https://github.com/python/cpython/blob/main/Lib/test/test_fstring.py. If we reuse that code, we have to verify the license and copyright. It may be easier to create our own?

Refs:

- https://docs.python.org/2/reference/lexical_analysis.html#string-literals
- https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals

**Comment by Felixoid** [comment]
It's just an opinion here, but Python 2 has been dead for over 5 years. Let the noble rest in peace.

**Comment by f4nb0y** [comment]
That's true, but folks are still maintaining legacy Python 2 codebases. Vim itself supported 16-bit MS-DOS until 2007 :-)

As long as we have https://github.com/vim/vim/blob/master/runtime/syntax/python2.vim in the source tree, I'd like to keep it functional. Especially the compatibility features which encourage incremental migration.

**Comment by f4nb0y** [comment]
I've opened #17726 for bytes, b-strings and u-strings.

**Comment by dkearns** [comment]
> We also need tests to ensure that we support as many cases as possible, and existing highlighting is not impacted. [@dkearns](https://github.com/dkearns) added a large test to [#14057](https://github.com/vim/vim/pull/14057), but some of that appears to be copied from https://github.com/python/cpython/blob/main/Lib/test/test_fstring.py. If we reuse that code, we have to verify the license and copyright. It may be easier to create our own?

https://github.com/python/cpython/blob/c7d24b81c376b0cf0b34f861cb18c1b1e4eac27b/LICENSE#L65C1-L67C33

> Starting with Python 3.8.6, examples, recipes, and other code in
> the documentation are dual licensed under the PSF License Version 2
> and the Zero-Clause BSD license.

The tests are PSF, I added a better header to the test file.

**Comment by dkearns** [comment]
> As long as we have https://github.com/vim/vim/blob/master/runtime/syntax/python2.vim in the source tree, I'd like to keep it functional. Especially the compatibility features which encourage incremental migration.

Yes, please.  There's no reason not to continue maintaining it while there's someone willing to do the work.

**Comment by zvezdan** [comment]
> It's just an opinion here, but Python 2 has been dead for over 5 years. Let the noble rest in peace.

I agree with @Felixoid!

I'd like to draw attention to the note in the comment at the top of _python2.vim_ syntax file.
At Bram's request, I added the file as a **historical record** for the benefit of the remaining Python 2 users, from the last commit that still supported correctly both Python 2 and 3 (up to 3.5).
I wanted to keep only Python 3 support and had a change for the main _python.vim_ syntax file that was breaking the compatibility with Python 2 (it was necessary), but Bram insisted on keeping at least some support for Python 2 users through this file. No further changes were intended to _ever_ be made in that file.

My opinion is that:

* the number of Python 2 users is dwindling
* none of them complained about the `u`, `b`, or `r` string highlight rendering since the split of syntax files happened
* maintaining a **historical record** file, such as _python2.vim_ with _after the fact_ corrections has diminishing returns
* instead, Python 2 users should be encouraged to _move on from the deprecated and unsupported language version_
  * to me this is a topic of a *language support*, not an *editor* support

**Comment by chrisbra** [comment]
Thanks. I understand your reasoning and it is causing a bit of a pain to maintain it. However, there are obviously still people that want to make further improvements to python2.vim for one reason or another. So if you do not want to maintain this anymore, you can pass the maintainer role to e.g. @f4nb0y if he wants to make further improvements there.

**Comment by zvezdan** [comment]
@chrisbra It's OK. If the majority feels we should update, I'm OK with that. There shouldn't be any frequent updates to the _python2.vim_ syntax file after this small highlight correction for Unicode escapes and strings.

We should just address the comments on the latest PR from @f4nb0y, see if he wants to merge the changes I've done in https://github.com/vim/vim/pull/17735, and add the change to the date of last update and to the note that indicates that there was a change after that last _historical record_ commit.

If @f4nb0y wants to commit only his PR, that's also OK. I can update the comments, note, ..., later.

**Comment by f4nb0y** [comment]
Thanks @zvezdan - I've addressed your comments in #17729.

The only other change I'm proposing for Python 2 is highlighting b-strings. This prefix was added in PEP 3112 to encourage migration to Python 3.

**Comment by Felixoid** [comment]
I am not sure why the issue is closed, the commit only adds something to the python-2, but not the f-strings

**Comment by zvezdan** [comment]
> The only other change I'm proposing for Python 2 is highlighting b-strings. This prefix was added in PEP 3112 to encourage migration to Python 3.

@f4nb0y Thanks. I'm OK with that if you want to add it. Bytes were added in Python 2.6. IIUC, that would only highlight the letter `b` together with the string (which is bytes in Python 2), so, it's probably not that important to Python 2 users and is probably rarely used in the existing code. However, if consistency is the goal, we could add it.

Do you want to prepare another PR? I see the tests for bytes are already in the previous PR and the dumps would probably need to be adjusted after the highlighting change.

**Comment by f4nb0y** [comment]
Done.

**Comment by jgb** [comment]
@glepnir this issue is about proper f-string syntax highlighting support, which isn't implemented yet, so I think it's too early to mark this as completed.

**Comment by glepnir** [comment]
Weird, I don't touch this issue before. reopend.. 

Edit: oh because i sync neovim commit then the commit msg mentioned this issue ...

**Comment by Felixoid** [comment]
And it's closed again

**Comment by chrisbra** [comment]
yes, because b7fc24d3a3d21ccf1461c703eb7ff07ef3994c54 introduced f-strings for python syntax highlighting

**Comment by A4-Tacks** [comment]
Is the expression in brace not highlighted?

current:

![Image](https://github.com/user-attachments/assets/3cdae8ee-bf04-4232-bccd-2b8be217558e)

expected:

![Image](https://github.com/user-attachments/assets/018303bd-f5ab-4c3c-8176-a8ad015b3f93)

**Comment by chrisbra** [comment]
ping @f4nb0y

**Comment by Felixoid** [comment]
<img width="813" height="454" alt="Image" src="https://github.com/user-attachments/assets/fd2e684e-d6d7-4c4e-972d-d43f32239071" />

Same here

**Comment by f4nb0y** [comment]
The next PR handles f-string replacement fields.

Working on it now :-)

**Comment by f4nb0y** [comment]
Opened draft PR #17784

Please confirm that it matches fields and delimiters in all cases, then we can discuss highlighting inside fields.

**Comment by Felixoid** [comment]
Thanks. Could you reopen the issues, please?

**Comment by f4nb0y** [comment]
Hi Folks,

#17962 is the final PR for this issue. Please confirm that it highlights all of your examples inside f-strings, and does not regress highlighting anywhere else.
